tools/create_all_possible_share_scenarios.py

```diff
@@ -0,0 +1,353 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# vim: expandtab shiftwidth=4 softtabstop=4
+#
+import sys
+
+try:
+    import owncloud
+except ImportError:
+    sys.stderr.writelines([
+        'This script requires pyocclient: https://github.com/owncloud/pyocclient\n',
+        'You can set PYTHONPATH to point to a local checkout of pyocclient using:\n',
+        '  export PYTHONPATH=/path/to/pyocclient\n'
+    ])
+    sys.exit(2)
+
+#
+# See https://github.com/owncloud/enterprise/issues/3701 for an explanation about the letters and tuples
+#
+# The following letters are used as symbols for the different share kind axes:
+#
+# - direction:
+#     - I: incoming
+#     - O: outgoing
+# - recipient type:
+#     - U: user, group, link, guest, federated share (for now).
+#     - L: link share. (there is no IL incoming link share)
+# - inheritance:
+#     - P: parent / direct share
+#     - C: child: share entry exists on one of the parents
+#
+# - possible combinations:
+#     - IUP: incoming user share on current resource
+#     - IUC: incoming user share exists on a parent resource
+#     - ILP: no incoming link shares possible
+#     - ILC: no incoming link shares possible
+#     - OUP: outgoing user share on current resource
+#     - OUC: outgoing user share exists on a parent resource
+#     - OLP: outgoing link share on current resource
+#     - OLC: outgoing link share exists on a parent resource
+#
+# - every resource can have multiple combinations of the above, without repetition:
+#     - for example: IUC, OUP, OLP means "incoming user share exists on a parent resource, the current resource has been shared by the current user to the outside with user share and link share"
+#     - we don't repeat a tuple of letters here because "U" or "L" mean that at least one share of the given type exists
+#     - maximum set that a resource can have: IUP, IUC, OUP, OUC, OLP, OLC
+#
+
+# TODO: get vars from command line arguments and/or files
+
+Config  = {
+    'owncloud_url': 'http://localhost/owncloud',
+    'owncloud_login': "admin",
+    'owncloud_password': "admin",
+    'users': [
+        ["alice", "Alice"],
+        ["bob", "Bob"],
+        ["charlie", "Charlie"],
+        ["dave", "Dave"],
+    ],
+    'users_password': 'test'
+}
+
+KIND_DESCRIPTIONS = {
+    'IUP': 'incoming user share on current resource',
+    'IUC': 'incoming user share exists on a parent resource',
+    'OUP': 'outgoing user share on current resource',
+    'OUC': 'outgoing user share exists on a parent resource',
+    'OLP': 'outgoing link share on current resource',
+    'OLC': 'outgoing link share exists on a parent resource',
```

Comment:
the parent and current resource descriptions are transposed.

Reply:
yeah, I know. P is the parent and C is the child